utils_cv/tracking/model.py

```diff
@@ -90,10 +95,107 @@ def write_video(results: Dict[int, List[TrackingBbox]],
-                    result_filename: str = 'results.txt') -> str:
+def plot_single_frame(
+    results: Dict[int, List[TrackingBbox]], input_video: str, frame_id: int
+) -> None:
+    """ 
+    Plot the bounding box and id on a wanted frame. Display as image to front end. 
+
+    Args:
+        results: dictionary mapping frame id to a list of predicted TrackingBboxes
+        input_video: path to the input video
+        frame_id: frame_id for frame to show tracking result
+    """
+    results = OrderedDict(sorted(results.items()))
+
+    # read video and initialize new tracking video
+    video = cv2.VideoCapture()
+    video.open(input_video)
+
+    image_width = int(video.get(cv2.CAP_PROP_FRAME_WIDTH))
+    image_height = int(video.get(cv2.CAP_PROP_FRAME_HEIGHT))
+    fourcc = cv2.VideoWriter_fourcc(*"MP4V")
+    frame_rate = int(video.get(cv2.CAP_PROP_FPS))
+
+    # assign bbox color per id
+    unique_ids = list(
+        set([bb.track_id for frame in results.values() for bb in frame])
+    )
+    color_map = assign_colors(unique_ids)
+
+    # Extract wanted frame:
+    video.set(cv2.CAP_PROP_POS_FRAMES, frame_id)
+    _, cur_image = video.read()
+
+    # Extract tracking results for wanted frame, and draw bboxes
+    cur_tracks = results[frame_id]
+    if len(cur_tracks) > 0:
+        cur_image = draw_boxes(cur_image, cur_tracks, color_map)
+
+    im = Image.fromarray(cur_image, "RGB")
+    b, g, r = im.split()
+    im = Image.merge("RGB", (r, g, b))
+    im.save("test_" + str(frame_id) + ".png")
+
+    IPython.display.display(im)
+
+
+def display_video(
+    results: Dict[int, List[TrackingBbox]], input_video: str
+) -> None:
+    """ 
+     Plot the predicted tracks on the input video. Displays to front-end as sequence of images stringed together in a video.
+
+    Args:
+        results: dictionary mapping frame id to a list of predicted TrackingBboxes
+        input_video: path to the input video        
+    """
+
+    results = OrderedDict(sorted(results.items()))
+
+    # assign bbox color per id
+    unique_ids = list(
+        set([bb.track_id for frame in results.values() for bb in frame])
+    )
+    color_map = assign_colors(unique_ids)
+
+    # read video and initialize new tracking video
+    video_reader = decord.VideoReader(input_video)
+
+    # set up ipython jupyter display
+    d_video = IPython.display.display("", display_id=1)
+
+    # Read each frame
+    frame_idx = 0
+    while True:
+        try:
+            cur_image = video_reader.next().asnumpy()
+
+            if len(cur_image.shape) != 3:
+                break
+            cur_tracks = results[frame_idx]
+            if len(cur_tracks) > 0:
+                cur_image = draw_boxes(cur_image, cur_tracks, color_map)
+
+            f = io.BytesIO()
+            im = Image.fromarray(cur_image)
+            im.save(f, "jpeg")
+
+            d_video.update(IPython.display.Image(data=f.getvalue()))
+            sleep(0.000001)
+            frame_idx += 1
```

Comment:
If you keep the while loop then consider moving this all the way up, since otherwise if the loop breaks the frame_idx is not in sync anymore with the results. That said, I hope we can get rid at the very least of the try..except block.

Reply:
I used the while and try block similar to action-recognition: https://github.com/microsoft/computervision-recipes/blob/staging/utils_cv/action_recognition/model.py 

Does this look better? We think it might be good to catch Exception for when the frame in the video is corrupted and giving errors. I can remove or edit if you think it's unnecessary
```
# Read each frame, add bbox+track_id, display frame    
    for frame_idx in range(len(results)): 
        try:
            frame_next=video_reader.next()
        except Exception as e:
            print('Error:', e)
            break
       
        im = frame_next.asnumpy()        
        cur_tracks = results[frame_idx]
        if len(cur_tracks) > 0:
            im = draw_boxes(im, cur_tracks, color_map)

        f = io.BytesIO()
        im = Image.fromarray(im)
        d_video.update(IPython.display.Image(data=f.getvalue()))
        sleep(0.000001)
```

Reply:
Looks much better!! If the code runs with our videos, I'd even remove the try..except statement to not "hide" a problem with videos

Reply:
Yes, it runs with our videos. Ok, will remove try...except.